### PR TITLE
[Merged by Bors] - feat: port data.nat.pairing

### PR DESCRIPTION
**Pull request by arienmalec**
Port of data.nat.pairing

### CODE REVIEW

**Comment by arienmalec** [comment]
Too: this as far as I could.

Remaining issues:

- Residual `rename_i`
- no tactic `change <foo> with <bar>`
- implicits in `unionᵢ_unpair_prod`
- mathport warnings

**Comment by antoinelab01** [comment]
I fixed the first three issues, but I'm not sure what the mathport warnings mean.

**Comment by antoinelab01** [comment]
I fixed the first three issues, but I'm not sure what the mathport warnings mean.

File: Mathlib/Data/Nat/Pairing.lean [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+theorem supᵢ_unpair {α} [CompleteLattice α] (f : ℕ → ℕ → α) :
+    (⨆ n : ℕ, f n.unpair.1 n.unpair.2) = ⨆ (i : ℕ) (j : ℕ), f i j := by
+  rw [← (supᵢ_prod : (⨆ i : ℕ × ℕ, f i.1 i.2) = _), ← Nat.surjective_unpair.supᵢ_comp]
+#align supr_unpair supᵢ_unpair
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
+theorem infi_unpair {α} [CompleteLattice α] (f : ℕ → ℕ → α) :
+    (⨅ n : ℕ, f n.unpair.1 n.unpair.2) = ⨅ (i : ℕ) (j : ℕ), f i j :=
+  supᵢ_unpair (show ℕ → ℕ → αᵒᵈ from f)
+#align infi_unpair infi_unpair
```
Comment (ChrisHughes24) [inline Mathlib/Data/Nat/Pairing.lean:196]:
```suggestion
theorem infᵢ_unpair {α} [CompleteLattice α] (f : ℕ → ℕ → α) :
    (⨅ n : ℕ, f n.unpair.1 n.unpair.2) = ⨅ (i : ℕ) (j : ℕ), f i j :=
  supᵢ_unpair (show ℕ → ℕ → αᵒᵈ from f)
#align infi_unpair infᵢ_unpair
```

File: Mathlib/Data/Nat/Pairing.lean [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+theorem Union_unpair {α} (f : ℕ → ℕ → Set α) :
+    (⋃ n : ℕ, f n.unpair.1 n.unpair.2) = ⋃ (i : ℕ) (j : ℕ), f i j :=
+  supᵢ_unpair f
+#align set.Union_unpair Set.Union_unpair
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
+theorem Inter_unpair {α} (f : ℕ → ℕ → Set α) :
+    (⋂ n : ℕ, f n.unpair.1 n.unpair.2) = ⋂ (i : ℕ) (j : ℕ), f i j :=
+  infi_unpair f
+#align set.Inter_unpair Set.Inter_unpair
```
Comment (ChrisHughes24) [inline Mathlib/Data/Nat/Pairing.lean:220]:
```suggestion
theorem unionᵢ_unpair {α} (f : ℕ → ℕ → Set α) :
    (⋃ n : ℕ, f n.unpair.1 n.unpair.2) = ⋃ (i : ℕ) (j : ℕ), f i j :=
  supᵢ_unpair f
#align set.Union_unpair Set.unionᵢ_unpair

/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
theorem interᵢ_unpair {α} (f : ℕ → ℕ → Set α) :
    (⋂ n : ℕ, f n.unpair.1 n.unpair.2) = ⋂ (i : ℕ) (j : ℕ), f i j :=
  infᵢ_unpair f
#align set.Inter_unpair Set.interᵢ_unpair
```

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**